File: test/ai.c
#include <stdio.h>
#include <stdlib.h>

// 判断是否为闰年
int isLeapYear(int year)
{
    return (year % 4 == 0 && year % 100 != 0) || (year % 400 == 0);
}

// 获取指定月份的天数
int getDaysInMonth(int month, int year)
{
    int days[] = {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
    if (month == 2 && isLeapYear(year))
    {
        return 29;
    }
    return days[month - 1];
}

// 获取指定日期是星期几（0 - 6，0 表示星期日）
int getDayOfWeek(int day, int month, int year)
{
    if (month < 3)
    {
        month += 12;
        year--;
    }
    int c = year / 100;
    year %= 100;
    int w = (c / 4 - 2 * c + year + year / 4 + 13 * (month + 1) / 5 + day - 1) % 7;
    return (w + 7) % 7;
}

// 打印单个月的日历标题（如“   一月 2025   ”，宽度 20）
void printMonthTitle(int month, int year)
{
    const char *monthNames[] = {
        "一月", "二月", "三月", "四月", "五月", "六月",
        "七月", "八月", "九月", "十月", "十一月", "十二月"};
    // 居中显示，总宽度 20
    printf("     %s %4d     ", monthNames[month - 1], year);
}

// 打印单个月的星期表头（宽度 20）
void printWeekHeader()
{
    printf(" 日 一 二 三 四 五 六  ");
}

// 打印从 startMonth 开始的 3 个月（一个季度）并排的日历
void printQuarter(int startMonth, int year)
{
    // 1. 标题行
    for (int i = 0; i < 3; i++)
    {
        printMonthTitle(startMonth + i, year);
        if (i < 2)
            printf("  "); // 月份间留两个空格
    }
    printf("\n");

    // 2. 星期表头行
    for (int i = 0; i < 3; i++)
    {
        printWeekHeader();
        if (i < 2)
            printf("  ");
    }
    printf("\n");

    // 3. 预先计算三个当月的首日偏移和天数
    int firstDay[3], daysInMon[3];
    for (int i = 0; i < 3; i++)
    {
        int m = startMonth + i;
        firstDay[i] = getDayOfWeek(1, m, year);
        daysInMon[i] = getDaysInMonth(m, year);
    }

    // 4. 按 6 行（星期）打印内容
    for (int week = 0; week < 6; week++)
    {
        for (int i = 0; i < 3; i++)
        {
            // 对当前月的这一周，遍历周日(0)到周六(6)
            for (int dow = 0; dow < 7; dow++)
            {
                int dayNum = week * 7 + dow - firstDay[i] + 1;
                if (dayNum >= 1 && dayNum <= daysInMon[i])
                {
                    printf("%3d", dayNum);
                }
                else
                {
                    printf("   ");
                }
            }
            if (i < 2)
                printf("  "); // 月份间留两个空格
        }
        printf("\n");
    }
}

int main()
{
    int year;
    printf("请输入要显示日历的年份: ");
    if (scanf("%d", &year) != 1)
    {
        fprintf(stderr, "输入无效！\n");
        return 1;
    }

    for (int start = 1; start <= 10; start += 3)
    {
        printQuarter(start, year);
        if (start < 10)
            printf("\n"); // 季度之间空一行
    }

    return 0;
}

File: test/ai.rs
use std::io;

// 判断是否为闰年
fn is_leap_year(year : i32) -> bool{
    (year % 4 == 0 && year % 100 != 0) || (year % 400 == 0)}

// 获取指定月份的天数
fn days_in_month(month : usize, year : i32)
    ->usize
{
    let days = [ 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31 ];
    if month
        == 2 && is_leap_year(year){
                    29} else
        {
            days[month - 1]
        }
}

// 获取指定日期是星期几（0 - 6，0 表示星期日）
fn day_of_week(day : i32, month : i32, year : i32) -> usize
{
    let(m, y) = if month < 3
    {
        (month + 12, year - 1)
    }
    else {
        (month, year)};
    let c = y / 100;
    let d = y % 100;
    let w = (c / 4 - 2 * c + d + d / 4 + (13 * (m + 1)) / 5 + day - 1) % 7;
    ((w + 7) % 7) as usize
}

// 打印单个月标题，宽度 20
fn print_month_title(month : usize, year : i32)
{
    let names = [
        "一月",
        "二月",
        "三月",
        "四月",
        "五月",
        "六月",
        "七月",
        "八月",
        "九月",
        "十月",
        "十一月",
        "十二月",
    ];
    let title = format !("{} {}", names[month - 1], year);
    let total_width = 20;
    let pad = (total_width - title.chars().count()) / 2;
    print !("{}{}{}", " ".repeat(pad), title, " ".repeat(total_width - pad - title.chars().count()));
}

// 打印星期表头
fn print_week_header()
{
    print !(" 日 一 二 三 四 五 六  ");
}

// 打印一行三个连续月
fn print_quarter(start_month : usize, year : i32)
{
    // 标题行
    for
        i in 0..3
        {
            print_month_title(start_month + i, year);
            if i
                < 2 { print !("  "); }
        }
    println !();

    // 星期表头
    for
        i in 0..3
        {
            print_week_header();
            if i
                < 2 { print !("  "); }
        }
    println !();

    // 计算首日与天数
    let mut first = [0; 3];
    let mut days = [0; 3];
    for
        i in 0..3
        {
            first[i] = day_of_week(1, (start_month + i) as i32, year);
            days[i] = days_in_month(start_month + i, year) as i32;
        }

    // 打印最多 6 行
    for
        week in 0..6
        {
        for
            i in 0..3
            {
            for
                dow in 0..7
                {
                    let day = week * 7 + dow as usize + 1 - first[i] as usize;
                    if day
                        >= 1 && (day as i32) <= days[i]
                        {
                            print !("{:>3}", day);
                        }
                    else
                    {
                        print !("   ");
                    }
                }
            if i
                < 2 { print !("  "); }
            }
        println !();
        }
}

fn main()
{
    let mut input = String::new ();
    println !("请输入要显示日历的年份: ");
    io::stdin().read_line(&mut input).unwrap();
    let year : i32 = input.trim().parse().expect("请输入一个有效的年份");

    for
        m in(1.. 12).step_by(3)
        {
            print_quarter(m, year);
            if m
                < 10 { println !(); }
        }
}